### PR TITLE
Add `powerPreference` preference #12533

### PR DESCRIPTION
**Pull request by luruke**

### CODE REVIEW

**Comment by luruke** [comment]
I had to move `webglcontextlost` and `webglcontextrestored` events listener *before* the creation of the gl context.

According to the spec:
> At WebGLRenderingContext creation time, this attribute will be ignored if no event listeners exist which would handle both "webglcontextlost" and "webglcontextrestored" events dispatched to the canvas element.

Honestly I had hard time to figure out any difference on my machine `MacBook Pro (13-inch, 2016) - 10.13.1, Intel Iris Graphics 540 1536 MB`.

I guess this parameter is taken into consideration for machine with multiple GPU?

File: src/renderers/WebGL2Renderer.js [hunk 1]
```diff
@@ -17,12 +17,13 @@ function WebGL2Renderer( parameters ) {
...
-		_antialias = parameters.antialias !== undefined ? parameters.antialias : false,
-		_premultipliedAlpha = parameters.premultipliedAlpha !== undefined ? parameters.premultipliedAlpha : true,
-		_preserveDrawingBuffer = parameters.preserveDrawingBuffer !== undefined ? parameters.preserveDrawingBuffer : false;
+    _alpha = parameters.alpha !== undefined ? parameters.alpha : false,
+    _depth = parameters.depth !== undefined ? parameters.depth : true,
+    _stencil = parameters.stencil !== undefined ? parameters.stencil : true,
+    _antialias = parameters.antialias !== undefined ? parameters.antialias : false,
+    _premultipliedAlpha = parameters.premultipliedAlpha !== undefined ? parameters.premultipliedAlpha : true,
+    _preserveDrawingBuffer = parameters.preserveDrawingBuffer !== undefined ? parameters.preserveDrawingBuffer : false,
+    _powerPreference = parameters.powerPreference !== undefined ? parameters.powerPreference : 'default';
```
Comment (Mugen87) [inline src/renderers/WebGL2Renderer.js]: Looks like some formatting issues...

**Review by Mugen87 — CHANGES_REQUESTED** [review]

File: src/renderers/WebGL2Renderer.js [hunk 2]
```diff
@@ -17,12 +17,13 @@ function WebGL2Renderer( parameters ) {
 	var _canvas = parameters.canvas !== undefined ? parameters.canvas : document.createElementNS( 'http://www.w3.org/1999/xhtml', 'canvas' ),
 		_context = parameters.context !== undefined ? parameters.context : null,
 
-		_alpha = parameters.alpha !== undefined ? parameters.alpha : false,
-		_depth = parameters.depth !== undefined ? parameters.depth : true,
-		_stencil = parameters.stencil !== undefined ? parameters.stencil : true,
-		_antialias = parameters.antialias !== undefined ? parameters.antialias : false,
-		_premultipliedAlpha = parameters.premultipliedAlpha !== undefined ? parameters.premultipliedAlpha : true,
-		_preserveDrawingBuffer = parameters.preserveDrawingBuffer !== undefined ? parameters.preserveDrawingBuffer : false;
+    		_alpha = parameters.alpha !== undefined ? parameters.alpha : false,
```
Comment (mrdoob) [inline src/renderers/WebGL2Renderer.js]: Seems like there is a mix of spaces and tabs for indentation in these lines...

File: src/renderers/WebGL2Renderer.js [hunk 3]
```diff
@@ -36,9 +37,13 @@ function WebGL2Renderer( parameters ) {
 			stencil: _stencil,
 			antialias: _antialias,
 			premultipliedAlpha: _premultipliedAlpha,
-			preserveDrawingBuffer: _preserveDrawingBuffer
+			preserveDrawingBuffer: _preserveDrawingBuffer,
+			powerPreference: _powerPreference,
```
Comment (mrdoob) [inline src/renderers/WebGL2Renderer.js]: The comma at the end is not needed.

File: src/renderers/WebGLRenderer.js [hunk 4]
```diff
@@ -199,9 +200,13 @@ function WebGLRenderer( parameters ) {
 			stencil: _stencil,
 			antialias: _antialias,
 			premultipliedAlpha: _premultipliedAlpha,
-			preserveDrawingBuffer: _preserveDrawingBuffer
+			preserveDrawingBuffer: _preserveDrawingBuffer,
+			powerPreference: _powerPreference,
```
Comment (mrdoob) [inline src/renderers/WebGLRenderer.js]: The comma at the end is not needed.

**Comment by mrdoob** [comment]
> I had to move `webglcontextlost` and `webglcontextrestored` events listener *before* the creation of the gl context.
>
> According to the spec:
>> At WebGLRenderingContext creation time, this attribute will be ignored if no event listeners exist which would handle both "webglcontextlost" and "webglcontextrestored" events dispatched to the canvas element.

I don't understand... Do you mind giving more context?

**Comment by luruke** [comment]
@mrdoob sorry for those silly indentations issues.

From the spec: https://www.khronos.org/registry/webgl/specs/latest/1.0/#5.2.1
What I understand is that you must handle `webglcontextlost` and `webglcontextrestored` before the creation of the context, otherwise the powerPreference is ignored.

A related issue:
https://github.com/KhronosGroup/WebGL/issues/2343

**Comment by luruke** [comment]
@mrdoob @Mugen87 
You can find more info here:
https://www.khronos.org/webgl/public-mailing-list/public_webgl/1703/msg00021.php

**Comment by Mugen87** [comment]
see #2343
and
> You MUST have a registered event handler for the webglcontextlost and webglcontextrestored events if you want the user agent to respect your request for high-performance.

@luruke So yes, i think the implementation is correct.

**Review by Mugen87 — APPROVED** [review]

**Comment by Mugen87** [comment]
Um, maybe we should add a comment. Something like:
`
// event listeners must be registered before WebGL Context is created, see #12753
`
Otherwise it's easy to overlooked this dependency...

**Comment by mrdoob** [comment]
Perfect!

**Comment by mrdoob** [comment]
Thanks!

**Pull request merged**